agents/self_evolving_car/train.py

```diff
@@ -0,0 +1,163 @@
+        self.bot_fitness = [0] * self.pop
+        self.fittest = None  # fittest object
+        self.mut_rate = 0.2  # mutation rate
+        self.distance_to_ball = [math.inf] * self.max_frames  # set high for easy minimum
+        self.input_formatter = LeviInputFormatter(team, index)
+        self.output_formatter = LeviOutputFormatter(index)
+
+    def initialize_agent(self):
+        self.reset()  # reset at start
+
+    def get_output(self, packet: GameTickPacket) -> SimpleControllerState:
+        # NEURAL NET INPUTS
+        inputs = self.input_formatter.create_input_array([packet])
+        inputs = [self.torch.from_numpy(x).float() for x in inputs]
+
+        my_car = packet.game_cars[self.index]
+        distance_to_ball_x = packet.game_ball.physics.location.x - my_car.physics.location.x
+        distance_to_ball_y = packet.game_ball.physics.location.y - my_car.physics.location.y
+        distance_to_ball_z = packet.game_ball.physics.location.z - my_car.physics.location.z
+        self.distance_to_ball[self.frame] = math.sqrt(distance_to_ball_x**2 + distance_to_ball_y**2 + distance_to_ball_z**2)
+
+        # RENDER RESULTS
+        action_display = "GEN: "+str(self.gen+1)+" | BOT: "+str(self.brain)#+" \nTURN: "+str(self.botList[self.brain].Nodes[1].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,4,5,6,7))+" \nBOOST: "+str(self.botList[self.brain].Nodes[2].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,8,9,10,11))+" \nTHROTTLE: "+str(self.botList[self.brain].Nodes(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,0,1,2,3))
+
+        draw_debug(self.renderer, my_car, packet.game_ball, action_display)
+
+        # # STOP EVOLVING WHEN THE BALL IS TOUCHED
+        # if packet.game_ball.latest_touch.player_name == "Self-Evolving-Car":
+        #     self.mut_rate = 0
+
+        # GAME STATE
+        car_state = CarState(boost_amount=100)
+        ball_state = BallState(Physics(velocity=Vector3(0, 0, 0), location=Vector3(0, -1000, 1200), angular_velocity=Vector3(0, 0, 0)))
+        game_state = GameState(ball=ball_state, cars={self.index: car_state})
+        self.set_game_state(game_state)
+
+        # NEURAL NET OUTPUTS
+        with self.torch.no_grad():
+            outputs = self.bot_list[self.brain].forward(*inputs)
+        self.controller_state = self.output_formatter.format_model_output(outputs, [packet])[0]
+
+        # KILL
+        if (my_car.physics.location.z < 100 or my_car.physics.location.z > 1950 or my_car.physics.location.x < -4000 or my_car.physics.location.x > 4000 or my_car.physics.location.y > 5000) and self.frame > 50:
+            #self.botList[self.brain].fitness = (1/self.frame+1)*100000
+            self.frame = self.max_frames
+
+        # LOOPS
+        self.frame = self.frame + 1
+
+        if self.frame >= self.max_frames:
+            self.frame = 0
+            self.calc_fitness()
```

Comment:
this part is fine in this file it would just call it on a separate object whose code is in a different file